### PR TITLE
[skip preview] setup action test

### PR DESCRIPTION
**Pull request by NiklasBuchfink**
General:
- [x] insert GITHUB_TOKEN via env
- [x] scope to PRs and file changes

Styling:
- [x] comment styling (markdown)
- [x] list missing messages

Test:
- [x] create comment
- [x] update comment
- [ ] release action
- [ ] adding to a new repo

Content:
- [ ] Branding
- [ ] Create a product page
- [ ] Extend Automation guide
- [ ] Update README

### CODE REVIEW

**Comment by samuelstroschein** [comment]
@NiklasBuchfink is the diffing so complicated? the github action provides better value if it diffs and only triggers if new lint reports come in. 

@janfjohannes interim workaround to trigger the github action when a project file changes substantially increases the barrier for adoption and risk for breaking the github action. If someone renamed an inlang project, the message.json file, adds a new project or whatever, the github action is broken.

**Comment by samuelstroschein** [comment]
seeing this message https://github.com/opral/monorepo/pull/2176#issuecomment-1952860075 increases the desire for diffing even more. yes the project file changed but this PR introduces no message updates. Hence, the message provides no value

**Comment by NiklasBuchfink** [comment]
> The view needs to be split by lint rule.

Ok, I can change that. I was not sure if this should be listed by errors/warnings or lint rules id

**Comment by felixhaeberle** [comment]
I would also suggest to:

- get rid of the warning ⚠️ icon, we have errors & warnings & in the future other lint rules. the best would be (also in regard to change control): 🛎️ Update translations

- give multiple links for each lint rule (scales better and lint rules can eventually be linked = what does this lint rule do?)

- naming fink prominently without explaining what the user can do hurts conversion: I would go for contribute (via Fink) as @samuelstroschein has proposed

**Comment by NiklasBuchfink** [comment]
> @NiklasBuchfink is the diffing so complicated?

@janfjohannes Therefore these two need to work:
1. Checkout a different branch (e.g. PR branch `github-lint-action` -> `main`)
2. Checkout a different repos (e.g. PR fork `NiklasBuchfink/monorepo` -> `opral/monorepo`)
Can you give me an ETA for this?

@samuelstroschein Would this approach be ok if we don't use the `paths` and instead:
1. use project path which is an input anyway
2. get the store location of the language files from the SDK (needs to be exposed) -> check if they are in the changedFiles of the PR (simple filter of file names)

**Comment by janfjohannes** [comment]
@NiklasBuchfink   if the paths filter is not an option and you cannot wait for lix to fix brnach switching, you could also switch branch with command line git,  then open the repo again with lix and compare the new output with the saved one.

**Comment by NiklasBuchfink** [comment]
> @NiklasBuchfink if the paths filter is not an option and you cannot wait for lix to fix brnach switching, you could also switch branch with command line git, then open the repo again with lix and compare the new output with the saved one.

@janfjohannes Can I execute the Git commands from the Typescript file similar to the CLI?

**Comment by samuelstroschein** [comment]
> @NiklasBuchfink  Can I execute the Git commands from the Typescript file similar to the CLI?

Yes, you can. Search for "node subprocess" or "node execute" 

@janfjohannes good workaround idea to call the git CLI!

**Comment by NiklasBuchfink** [comment]
> Yes, you can. Search for "node subprocess" or "node execute"

Ok, I will try this tomorrow 👍

**Comment by NiklasBuchfink** [comment]
### Quick Update:
- [x] comment is updated
- [x] diff works for branches using Git CLI
- [ ] diff works for forks using Git CLI (to be tested)
- [ ] multiple projects (comment must include project path/id to update the correct comment)

**Comment by samuelstroschein** [comment]
>  multiple projects (comment must include project path/id to update the correct comment)

Having the comment point to the project is a good idea! That avoids config for the user.

**Comment by NiklasBuchfink** [comment]
The last issue is the following:
If a PR comes from a fork, the permissions of the GITHUB_TOKEN are set to read only. This way I cannot create a comment. This is a [known issue](https://stackoverflow.com/questions/76952023/how-to-make-github-actions-safely-access-secrets-for-prs-created-from-forks).

Everything else works now.

**Comment by samuelstroschein** [comment]
@NiklasBuchfink let's ignore forks then? I assume commenting still work for pull requests opened on a fork to the upstream repo?

**Comment by NiklasBuchfink** [comment]
@samuelstroschein The check works now (See second action PR from NiklasBuchfink/monorepo-copy), but the commenting is the issue. We need the GitHub token to comment. For the CLA workflow it works somehow with the same token. 🤔

**Comment by NiklasBuchfink** [comment]
**Update**
- [x] commenting works for PRs from forks and branches
- [x] automatically checks all projects without the need for an input
- [x] lists results of all projects in one comment
- [x] debug logs in action in case some of the projects have errors (Skip project ... because of errors)
- [x] diffing if a project went from "no errors" to "some errors" and provide an error message:
image

**Comment by NiklasBuchfink** [comment]
PR from fork: https://github.com/opral/monorepo/pull/2291
PR from branch: https://github.com/opral/monorepo/pull/2297

**Comment by NiklasBuchfink** [comment]
Related issues to make this even better:
- https://github.com/opral/monorepo/issues/1127 (required to only show relevant messages/ filter multiple IDs)
- https://github.com/opral/monorepo/issues/1841 (show the correct number of messages)

**Comment by samuelstroschein** [comment]
@NiklasBuchfink should I review this? The PR is marked as draft (not ready for review)

**Comment by NiklasBuchfink** [comment]
@samuelstroschein I will implement the proposed changes before you review this.

I would argue for not listing all project errors in the comment, as it would get very messy with 6 errors like this:
image

Maybe I can just show the error type and a link to Fink where the details can be viewed. What do you think?

**Comment by samuelstroschein** [comment]
@NiklasBuchfink have you considered showing all errors but behind an accordion https://gist.github.com/pierrejoubert73/902cc94d79424356a8d20be2b382e1ab ?

**Comment by NiklasBuchfink** [comment]
@samuelstroschein Good idea, I can do it that way 👍

File: module.js [hunk 1]
```diff
@@ -0,0 +1,35 @@
+// mathUtils.js
+
```
Comment (janfjohannes) [inline module.js:2]: whats this file?

Comment (NiklasBuchfink) [reply module.js:2]: This was a test module during the module import problem. I forgot to delete it 👍

File: inlang/source-code/github-lint-action/src/main.ts [hunk 2]
```diff
@@ -1,97 +1,388 @@
...
-			deletions: 0,
-			changes: 0,
+		// Prepare head repo
+		let repoHead
+		if (isFork) {
+			core.debug("Fork detected, cloning base repository")
+			process.chdir("../../../")
+			await cloneRepository(headMeta)
+			process.chdir(headMeta.repo)
+			repoHead = await openRepository(process.cwd(), {
```
Comment (janfjohannes) [inline inlang/source-code/github-lint-action/src/main.ts:86]: see above

File: inlang/source-code/github-lint-action/src/main.ts [hunk 3]
```diff
@@ -1,97 +1,388 @@
...
+		core.debug(`Is fork: ${isFork}`)
 
-		let diffData = {
-			additions: 0,
-			deletions: 0,
-			changes: 0,
+		// Prepare head repo
+		let repoHead
+		if (isFork) {
+			core.debug("Fork detected, cloning base repository")
```
Comment (janfjohannes) [inline inlang/source-code/github-lint-action/src/main.ts:82]: why would we handle forks here?

File: inlang/source-code/github-lint-action/src/main.ts [hunk 4]
```diff
@@ -1,97 +1,388 @@
...
+			process.chdir("../../../")
+			await cloneRepository(headMeta)
+			process.chdir(headMeta.repo)
+			repoHead = await openRepository(process.cwd(), {
+				nodeishFs: fs,
+			})
+		} else {
+			core.debug("Fork not detected, fetching and checking out base repository")
```
Comment (janfjohannes) [inline inlang/source-code/github-lint-action/src/main.ts:90]: not sure why this is needed aat all, the pr will run on the branch after the merge simulated allrady by default in github.

File: inlang/source-code/github-lint-action/src/main.ts [hunk 5]
```diff
@@ -1,97 +1,388 @@
...
+			await fetchBranch(headMeta.branch)
+			await checkoutBranch(headMeta.branch)
+			await pull()
+			repoHead = await openRepository(process.cwd(), {
```
Comment (janfjohannes) [inline inlang/source-code/github-lint-action/src/main.ts:94]: see above

File: inlang/source-code/github-lint-action/src/main.ts [hunk 6]
```diff
@@ -1,97 +1,388 @@
...
+			}
+			core.debug(`stdout: ${stdout}`)
+			core.debug(`stderr: ${stderr}`)
+			resolve()
+		})
+	})
+}
+
+// Function to clone the base repository
+async function cloneRepository(repoData: { link: string; branch: string }) {
```
Comment (janfjohannes) [inline inlang/source-code/github-lint-action/src/main.ts:370]: not sure why needed, this is allready in the yaml as one of the steps before this code runs, no?

**Review by janfjohannes — CHANGES_REQUESTED** [review]

File: inlang/source-code/github-lint-action/src/index.test.ts [hunk 7]
```diff
@@ -1,6 +1,102 @@
...
+					},
+				],
+			},
+		],
+	},
+]
+
+async function setupProject(enabledLintRule?: MessageLintRule) {
+	const repo = await mockRepo()
+	await fs.mkdir("project.inlang", { recursive: true })
```
[the remark this answers is not in the document]
Comment (NiklasBuchfink) [reply inlang/source-code/github-lint-action/src/index.test.ts:62]: The test doesn't work at all. I must split up the action file and write tests for all functions. I can't run the whole action at once, as it relies heavily on GitHub APIs. So, testing is missing completely atm.

File: inlang/source-code/github-lint-action/src/main.ts [hunk 8]
```diff
@@ -1,97 +1,388 @@
...
+		const token = process.env.GITHUB_TOKEN
+		if (!token) {
+			throw new Error("GITHUB_TOKEN is not set")
+		}
+		const { owner, repo } = github.context.repo
+		const prNumber = github.context.payload.pull_request?.number
 
-		// Debug logs are only output if the `ACTIONS_STEP_DEBUG` secret is true
-		core.debug(`I got all inputs: ${owner} ${repo} ${pr_number} ${token}`)
+		const repoBase = await openRepository(process.cwd(), {
```
[the remark this answers is not in the document]
Comment (NiklasBuchfink) [reply inlang/source-code/github-lint-action/src/main.ts]: Done ✅

File: inlang/source-code/github-lint-action/src/index.test.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (NiklasBuchfink) [reply inlang/source-code/github-lint-action/src/index.test.ts]: I deleted the test file as it makes no sense to have something broken in there.

File: inlang/source-code/github-lint-action/src/index.test.ts [hunk 9]
```diff
@@ -1,6 +1,102 @@
...
+	const _import = async (name: string) => {
+		if (name === "lintRule.js") {
+			return { default: enabledLintRule as MessageLintRule } satisfies InlangModule
+		}
+		return {
+			default: _mockPlugin,
+		}
+	}
+
+	return await loadProject({
```
[the remark this answers is not in the document]
Comment (NiklasBuchfink) [reply inlang/source-code/github-lint-action/src/index.test.ts]: see above

File: inlang/source-code/github-lint-action/src/main.ts (hunk 3 above)
Comment (NiklasBuchfink) [reply inlang/source-code/github-lint-action/src/main.ts]: The debug message was irritating. We check out the base repo first and need the head repo of a fork or the head branch to compare both. I updated the comment and we talked about it ✅

**Pull request merged**